### PR TITLE
language: upgrades: no generics for data types having the instances already

### PR DESCRIPTION
**Pull request by robin-da**
We only derive generic instances on the fly for data types that don't
have them already.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
This just doesn't derive them again for those that clearly have them already (by looking at the source). Overlappable instances might work, but that would mean we ignore custom made generic instances. This does not work for orphan instances, but if there are orphan instances it's maybe better to fail and let the user decide which instance should be used.

**Comment by neil-da** [comment]
I'd be inclined to follow the simpler path of turning on orphan instances and just ignoring pre-written instances. Is writing custom instances a valid or sensible thing to do?

**Comment by robin-da** [comment]
I tested turning overlappable instances on, but I still get a 'duplicate instance' error.

**Comment by cocreature** [comment]
Overlappable instances only work if one is more specific than the other.

**Comment by robin-da** [comment]
Once we start reading the dalf instead of the source we most likely bump into this problem again. If 'overlapping instances' would work that would be way easier. Any suggestions @cocreature @neil-da ?

File: daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/Upgrade.hs [hunk 1]
```diff
@@ -38,8 +38,18 @@ generateGenInstancesModule qual (pkg, L _l src) =
             tcdTyVars
             tcdDataDefn
         | L _ (TyClD _x DataDecl {..}) <- hsmodDecls src
+        , not $ hasGenDerivation tcdDataDefn
         ]
 
+    hasGenDerivation :: HsDataDefn GhcPs -> Bool
+    hasGenDerivation HsDataDefn {..} =
+        or [ name `elem` [nameOccName n | n <- genericClassNames]
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/Upgrade.hs]:
```suggestion
        or [ name `elem` map nameOccName genericClassNames
```

File: daml-assistant/integration-tests/src/Main.hs [hunk 2]
```diff
@@ -196,9 +196,12 @@ packagingTests tmpDir = testGroup "packaging"
         step "Creating project a..."
         createDirectoryIfMissing True (projectA </> "daml")
         writeFileUTF8 (projectA </> "daml" </> "Main.daml") $ unlines
-            [ "daml 1.2"
+            [ "{-# LANGUAGE EmptyCase #-}"
```
Comment (hurryabit) [inline daml-assistant/integration-tests/src/Main.hs]: Why do we need that now?

Comment (robin-da) [reply daml-assistant/integration-tests/src/Main.hs]: Because the generic instance derivation apparently creates an empty case. I guess for unit.

**Comment by robin-da** [comment]
So I think the problem with overlapping instances will actually go away once we start creating the upgrade code from dalfs instead of the source. I'll merge this one, as it makes the current solution more usable and reiterate.

**Pull request merged**